File: recursion/javascript/solutions/inits.js
/**
 * Solution for inits.
 */

const { EMPTY_LIST, Node, isEmpty, prepend, unprepend, toArray } = require('./linkedList');
const { mapList } = require('./mapList');

function inits(list) {
  if (isEmpty(list)) {
    return prepend(EMPTY_LIST, EMPTY_LIST);
  }

  let [first, rest] = unprepend(list);

  return prepend(EMPTY_LIST, mapList((prefix) => prepend(first, prefix), inits(rest)));
}

if (require.main === module) {
  let result = inits(Node.fromArray([1, 2, 3]));
  let initsList = toArray(result).map((t) => (t === null ? [] : toArray(t)));
  console.assert(JSON.stringify(initsList) === '[[],[1],[1,2],[1,2,3]]');

  let emptyResult = inits(EMPTY_LIST);
  let emptyInits = toArray(emptyResult).map((t) => (t === null ? [] : toArray(t)));
  console.assert(JSON.stringify(emptyInits) === '[[]]');

  console.log('All tests passed!');
}

module.exports = { inits };

File: recursion/python/solutions/inits.py
"""Solution for inits."""

from linked_list import EMPTY_LIST, Node, is_empty, prepend, unprepend, to_python_list
from map_list import map_list


def inits(lst):
    """
    Return list of all prefixes (inits).

    inits(1 -> 2 -> 3) => () -> (1) -> (1 -> 2) -> (1 -> 2 -> 3)
    inits(()) => ()
    """
    if is_empty(lst):
        return prepend(EMPTY_LIST, EMPTY_LIST)

    first, rest = unprepend(lst)

    return prepend(EMPTY_LIST, map_list(lambda tail: prepend(first, tail), inits(rest)))


if __name__ == '__main__':
    # Test cases
    result = inits(Node.from_list([1, 2, 3]))
    result_list = to_python_list(result)
    assert to_python_list(result_list[0]) == []
    assert to_python_list(result_list[1]) == [1]
    assert to_python_list(result_list[2]) == [1, 2]
    assert to_python_list(result_list[3]) == [1, 2, 3]

    result = inits(EMPTY_LIST)
    result_list = to_python_list(result)
    assert len(result_list) == 1
    assert to_python_list(result_list[0]) == []

    print("All tests passed!")
